### PR TITLE
#655 reflect CRUD resource errors in deployed status condition

### PR DESCRIPTION
**Pull request by gtully**

### CODE REVIEW

File: api/v1beta1/activemqartemis_types.go [hunk 1]
```diff
@@ -685,6 +685,7 @@ const (
 	DeployedConditionNotReadyReason         = "PodsNotReady"
 	DeployedConditionZeroSizeReason         = "ZeroSizeDeployment"
 	DeployedConditionValidationFailedReason = "ValidationFailed"
+	DeployedConditionResourceErrorReason    = "ResourceError"
```
Comment (gtully) [inline api/v1beta1/activemqartemis_types.go:688]: would ResourceCRUDerror be better?

File: controllers/activemqartemis_controller.go [hunk 2]
```diff
@@ -160,12 +160,12 @@ func (r *ActiveMQArtemisReconciler) Reconcile(ctx context.Context, request ctrl.
 
 	if valid, result = validate(customResource, r.Client, r.Scheme, *namer); valid {
 
-		reconciler.Process(customResource, *namer, r.Client, r.Scheme)
+		err = reconciler.Process(customResource, *namer, r.Client, r.Scheme)
```
[the remark this answers is not in the document]
Comment (brusdev) [reply controllers/activemqartemis_controller.go:163]: If we want different type of conditions we could add new error implementations and the caller can then use a [type assertion](https://go.dev/doc/go_spec.html#Type_assertions) to identify the source error type.

File: controllers/activemqartemis_reconciler.go [hunk 3]
```diff
@@ -1152,22 +1154,37 @@ func (reconciler *ActiveMQArtemisReconcilerImpl) ProcessResources(customResource
...
 	//empty the collected objects
 	reconciler.requestedResources = nil
 
-	return nil
+	if len(compositeError) == 0 {
+		return nil
+	} else {
+		// maybe errors.Join in go1.20
+		// using %q(uote) to keep errors separate
+		return fmt.Errorf("%q", compositeError)
```
Comment (brusdev) [inline controllers/activemqartemis_reconciler.go:1177]: The function [errors.Join](https://pkg.go.dev/errors@master#Join) could be used to create an error that wraps the given errors. The error formats as the concatenation of the strings obtained by calling the Error method of each element of errs, with a newline between each string.

Comment (gtully) [reply controllers/activemqartemis_reconciler.go:1177]: not sure about having new lines in a condition message, but also join is in go1.20

File: api/v1beta1/activemqartemis_types.go (hunk 1 above)
Comment (brusdev) [reply api/v1beta1/activemqartemis_types.go:688]: I vote for `ResourceError` or `DeploymentResourceError`

**Review by brusdev — COMMENTED** [review]
This is a great enhancement!!! It will allows to check deployment resource errors from the CR.

**Review by brusdev — APPROVED** [review]

File: controllers/activemqartemis_reconciler.go (hunk 3 above)
Comment (brusdev) [reply controllers/activemqartemis_reconciler.go:1177]: A newline between each string could be a good separator but I don't know if the Kubernetes/OpenShift console renders the new lines in a condition message.

File: api/v1beta1/activemqartemis_types.go (hunk 1 above)
Comment (gtully) [reply api/v1beta1/activemqartemis_types.go:688]: now thinking: KindCRUDError or is that more incomprehensible? it is more accurate

Comment (gtully) [reply api/v1beta1/activemqartemis_types.go:688]: resource means cpu/memory etc in kube, a type is a kind, and we are doing CRUD as part of the reconcile loop.

File: controllers/activemqartemis_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gtully) [reply controllers/activemqartemis_controller.go:163]: in this case, the message is the error, the crud errored. I don't think we want conditions per error. These errors should only be transient or real and rare.

File: api/v1beta1/activemqartemis_types.go (hunk 1 above)
Comment (gtully) [reply api/v1beta1/activemqartemis_types.go]: hmm, the crud acronym is too much, going to revert to the original, resource is more general than CPU and RAM and is suitable.

**Pull request merged**